moby/moby
# daemon: allow shimv2 runtimes to be configured

- Relates to #45030
- Relates to https://github.com/Mirantis/cri-dockerd/issues/111#issuecomment-1434253655

Kubernetes only permits RuntimeClass values which are valid lowercase RFC 1123 labels, which disallows the period character. This prevents cri-dockerd from being able to support configuring alternative shimv2 runtimes for a pod as shimv2 runtime names must contain at least one period character. 


**- What I did**

- Added support for configuring named shimv2 runtimes in daemon.json so that runtime names can be aliased to Kubernetes-compatible names.
- Added support for setting options for shimv2 runtimes in daemon.json.

**- How I did it**

I took heavy inspiration [from cri-containerd configuration](https://github.com/containerd/containerd/blob/main/docs/cri/config.md#runtime-classes) and [its implementation](https://github.com/containerd/containerd/blob/2456e983eb9e37e47538f59ea18f2043c9a73640/pkg/cri/server/helpers.go#L336-L374). The names of the new daemon runtime config fields have been selected to correspond with the equivalent field names in cri-containerd's configuration so that users can more easily follow documentation from the runtime vendor written for cri-containerd and apply it to daemon.json.

**- How to verify it**

Create a `/etc/docker/daemon.json` containing
```json
{
	"runtimes": {
		"custom": {
			"runtime-type": "io.containerd.runc.v2",
			"options": {
			       "BinaryName": "/usr/local/bin/doesnotexist"
			}
		}
	}
}
```
and (re)start the daemon. Try running a container with that runtime and examine the error message to see if the `BinaryName` option was respected.
```console
$ docker run --rm --runtime custom alpine ls
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: unable to retrieve OCI runtime error (open /var/run/docker/containerd/daemon/io.containerd.runtime.v2.task/moby/0c07ce57ea72bc26d0ffa0d1501506329fb15054e56072c076f0efc56f7eb342/log.json: no such file or directory): fork/exec /usr/local/bin/doesnotexist: no such file or directory: unknown.
```

**- Description for the changelog**

- Alternative runtimes can now be configured in daemon.json, enabling runtime names to be aliased and options to be passed.

**- A picture of a cute animal (not mandatory but encouraged)**
